File: zenoh_pc_nodes/cpp/src/zenoh_test_sub.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/std_msgs/z_Int32.h>
#include <iostream>

class CounterSubscriberNode : public ZenohNode {
public:
    CounterSubscriberNode() : ZenohNode("counter_subscriber") {
        std::cout << "[Node] " << this->z_get_name() << " has been started\n";

        // Create subscription to standard ROS 2 message type (Int32) and depth 10
        sub_ = this->z_create_subscription<z_Int32>(
            "robot/sim_counter",
            [this](const z_Int32& msg) -> void {
                this->listener_callback(msg);
            },
            10
        );
    }

    ~CounterSubscriberNode() {
        if (sub_) {
            delete sub_;
        }
        std::cout << "Destroying the Zenoh Node...\n";
    }

private:
    ZenohSubscription<z_Int32>* sub_ = nullptr;

    void listener_callback(const z_Int32& msg) {
        std::cout << "[RECV FROM MCU] sim_counter: " << msg.data << "\n";
    }
};

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    // Initialize Zenoh session with config object
    ZenohConfig config;
    config.host = "10.42.0.50";
    config.port = 7447;

    if (!ZenohNode::init(config)) {
        return -1;
    }

    {
        CounterSubscriberNode node;
        node.z_spin();
    } // Node goes out of scope and destroys subscriptions cleanly before session closes via atexit

    return 0;
}

File: zenoh_pc_nodes/python/zenoh_test_sub.py
#!/usr/bin/env python3
import sys
import os

# Ensure the shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.std_msgs import z_Int32

class CounterSubscriberNode(ZenohNode):
    def __init__(self) -> None:
        super().__init__("counter_subscriber")
        print(f"[Node] {self.z_get_name()} has been started")

        # Create subscription with standard ROS 2 message type (Int32) and depth 10
        self.sub = self.z_create_subscription(
            z_Int32,
            "robot/sim_counter",
            self.listener_callback,
            10
        )

    def listener_callback(self, msg: z_Int32) -> None:
        # Message is already deserialized into standard z_Int32 class!
        print(f"[RECV FROM ESP32-S3] sim_counter: {msg.data}")


def main() -> None:
    cfg = ZenohConfig(host="10.42.0.50", port=7447)
    if not ZenohNode.init(cfg):
        return
    node_instance = None
    try:
        node_instance = CounterSubscriberNode()
        node_instance.z_spin()

    except KeyboardInterrupt:
        print("\n[CTRL+C]>>> Interrupted by the User.")
        
    except Exception as e:
        print(f"Critical Error: {e}")
        
    finally:
        if node_instance is not None:
            print("Destroying the Zenoh Node...")
            node_instance.z_destroy()


if __name__ == '__main__':
    main()